Add support for user-configurable workflows

SamVerschueren opened the pull request:
This PR fixes #36 by adding support for user configurable workspaces. It's as simple as creating a `config.json` file in your workflow package and it will open it with the default text editor when installing the workflow.

SamVerschueren commented on index.js:
```diff
@@ -99,6 +100,10 @@ alfy.config = new Conf({
 });
 
+alfy.userConfig = new AlfredConfig({
+	cwd: alfy.alfred.data
```
This is the default. I wasn't sure if we should pass it through but might be good to be explicit? Could keep it with just `new AlfredConfig()` though.

sindresorhus replied:
If it's the default in `AlfredConfig`, I don't really see the need to specify it here.

SamVerschueren commented on readme.md:
````diff
@@ -378,6 +378,19 @@ alfy.config.get('unicorn');
 //=> '🦄'
 ```
 
+#### userConfig
````
Not sure if we have to document how this works? By creating a `config.json` file in the root of your package? Or link directly to https://github.com/SamVerschueren/alfred-config#workflow-configuration?

sindresorhus replied:
I would add a short description of how it works, and then link to alfred-config for more details. I would also include the use-cases here. Not clear from the property what a Alfred workflow creator would use it for when they read the docs.

https://github.com/SamVerschueren/alfred-config#workflow-configuration should describe where to put the JSON file, etc.

sindresorhus commented on readme.md:
```diff
@@ -378,6 +378,19 @@ alfy.config.get('unicorn');
 
+#### userConfig
+
+Type: `object`
```
```suggestion
Type: `Map`
```

sindresorhus commented on readme.md:
```diff
@@ -378,6 +378,19 @@ alfy.config.get('unicorn');
+
+Type: `object`
+
+If the workflow supports user configuration, it will export a readable [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) with the user workflow configuration.
```
Shouldn't it **always** export the `Map`? Only it will be empty if the workflow doesn't support user config. It's also not clear what it means for a workflow to "support user configuration".

SamVerschueren replied:
You are right, it seems that it does export a `Map` always.

sindresorhus commented:
Awesome how few changes `index.js` needed :)

sindresorhus commented:
@SamVerschueren Bump, in case you missed it. If you're just busy, feel free to ignore.

SamVerschueren commented:
Should definitely finish this one :). Thanks for the reminder.

sindresorhus commented:
@SamVerschueren Bump :)

sindresorhus commented:
Closing for lack of activity.

SamVerschueren commented:
I feel so bad for letting this one go stale. I think we are super close though.

I addressed your feedback. I'm going to upgrade the dependencies of `alfred-config` and clean it up a bit as well.

SamVerschueren commented:
Tests are failing, waiting for #135.

@sindresorhus could you check that the docs are better? It's been a while since I worked on this. I think it would be nice to have a link to a Alfy workflow which uses this. I could for instance setup a `alfy-example` package or something which has this and could work as a showcase.

I'm thinking about a workflow which shows you all the pokemon, and allows you to specify the levenstein distance for fuzzy search. If you have anything better, I'm up for it 😂 .

SamVerschueren commented on index.js:
```diff
@@ -99,6 +100,8 @@ alfy.config = new Conf({
 	cwd: alfy.alfred.data
 });
 
+alfy.userConfig = new AlfredConfig();
```
Ok, so apparently I have to fix something upstream first. Because I do `path.dirname(getEnv())` instead of `getEnv()` 🤦 .

https://github.com/SamVerschueren/alfred-config/blob/master/index.js#L11

SamVerschueren replied:
Done

SamVerschueren commented:
Fixed `alfred-config` and tested it locally if it still works as expected :p. Seems to work fine :).

sindresorhus commented:
> I feel so bad for letting this one go stale.

Don't worry. I was just trying to clean up my PR list.

sindresorhus commented:
I agree it would be useful to link to an example. It's not entirely clear still how you would set up everything.

> I'm thinking about a workflow which shows you all the pokemon, and allows you to specify the levenstein distance for fuzzy search. If you have anything better, I'm up for it 😂 .

I like the idea of Pokemon list, but I would make the language configurable.

sindresorhus commented:
Thanks for finishing this :)

SamVerschueren commented:
Oh configurable language is a good one 🎉
